[WebGL] Abstract working with EGL sync objects created from MTLSharedEvent

djg opened the pull request:
#### 023f54b8e5b80830c6d4eee7f54143aa4d15b9b9
<pre>
[WebGL] Abstract working with EGL sync objects created from MTLSharedEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=256954">https://bugs.webkit.org/show_bug.cgi?id=256954</a>
rdar://problem/109502752

Reviewed by Kimmo Kinnunen.

When the GPU process is enabled, the Web process no long has access to the APIs
required to create shared events passed from the UI process. This change added
APIs to allow creationg of sync objects from events that can be represented by a
MachSendRight (MTLSharedEvent), that can be used in the Web or GPU process.

Support work for WebXR in GPUP.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createEGLSync):
(WebCore::GraphicsContextGLANGLE::destroyEGLSync):
(WebCore::GraphicsContextGLANGLE::clientWaitEGLSyncWithFlush):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createEGLSync):
(WebCore::GraphicsContextGLCocoa::insertFinishedSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::createSyncWithSharedEvent): Deleted.
(WebCore::GraphicsContextGLCocoa::destroySync): Deleted.
(WebCore::GraphicsContextGLCocoa::clientWaitSyncWithFlush): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::createEGLSync):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(destroyEGLSync):
(clientWaitEGLSyncWithFlush):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLSync):
(WebKit::RemoteGraphicsContextGLProxy::clientWaitEGLSyncWithFlush):
* Tools/Scripts/generate-gpup-webgl:

Canonical link: <a href="https://commits.webkit.org/264659@main">https://commits.webkit.org/264659@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e66ef1430c6de4ced0b836f60e583bd875251f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9966 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8363 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10579 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8502 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11233 "100 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8444 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9502 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10110 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/8432 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7593 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15154 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7733 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11080 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7912 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8203 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6680 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8471 "Built successfully") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/8254 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7493 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7505 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2020 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1991 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11702 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8696 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7946 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2145 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1537,6 +1537,12 @@ class GraphicsContextGL : public RefCounted<GraphicsContextGL> {
     GraphicsContextGLAttributes contextAttributes() const { return m_attrs; }
     void setContextAttributes(const GraphicsContextGLAttributes& attrs) { m_attrs = attrs; }
 
+#if PLATFORM(COCOA)
```
GraphicsContextGL has already following:
- GCGLsync 
- GraphicsContextGL::fenceSync
- GraphicsContextGL::clientWaitSync
- GraphicsContextGL::deleteSync

These are OpenGL context sync objects (GLsync).

We cannot add the functions you're adding below, they refer to EGLSync objects, but overload the "sync" term and use incorrect return type.

So the fix needs to be either:
a) disambiguate GLsync vs EGLSync
b) only use EGLSync

Also:
you could try to add these too the autogenerated section.

```
#if PLATFORM(COCOA)
    using ExternalEGLSyncEvent = MachSendRight;
#else
   using ExternalEGLSyncEvent = int;
#endif
    virtual GCGLEGLSync createEGLSync(ExternalEGLSyncEvent, uint64_t) = 0;
    virtual void destroyEGLSync(GCGLEGLSync) = 0;
    virtual bool clientWaitEGLSyncWithFlush(GCGLEGLsync, uint64_t) = 0;
    // ========== Other functions.
 
 #if PLATFORM(COCOA)
    // Short term support for in-process WebGL.
    virtual GCGLEGLSync createEGLSync(id sharedSyncEvent, uint64_t) = 0;
#endif
   
```

djg replied:
`GCGLEGLSync` feels like prefix overload. Is `GCEGLSync` acceptable?

kkinnunen-apple replied:
good

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -233,7 +233,7 @@ void WebXROpaqueFramebuffer::endFrame()
     if (m_completionEvent) {
         auto gCGL = static_cast<GraphicsContextGLCocoa*>(&gl);
-        auto completionSync = gCGL->createSyncWithSharedEvent(m_completionEvent, m_renderingFrameIndex);
+        auto completionSync = gCGL->createSyncWithSharedEvent(m_completionEvent.get(), m_renderingFrameIndex);
```
Again, since you're abstacting something it'd be good to use the abstaction, e.g. work until you get to a point where you can use
```
gl.createEGLSync(m_completionEvent.get(),  m_renderingFrameIndex);

djg replied:
I have the use of the change in a follow up patch, but I can move those changes here too.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
```diff
@@ -71,6 +71,7 @@ class RemoteGraphicsContextGLProxy
     void didReceiveInvalidMessage(IPC::Connection&, IPC::MessageName) final { }
 
     // WebCore::GraphicsContextGL overrides.
+    GCEGLSync createEGLSync(ExternalEGLSyncEvent) final;
```
The main purpose of ExternalEGLSyncEvent is that we don't need platform-specific ifdefs for the arguments.
The main purpose we want that is that we then don't need the explicit manual function, rather we can just list this function in the autogen section and get the autogenerated implementation.

kkinnunen-apple commented on Tools/Scripts/generate-gpup-webgl:
```diff
@@ -124,6 +124,9 @@ messages -> RemoteGraphicsContextGL NotRefCounted Stream {{
     void MultiDrawArraysInstancedBaseInstanceANGLE(uint32_t mode, IPC::ArrayReferenceTuple<int32_t, int32_t, int32_t, uint32_t> firstsCountsInstanceCountsAndBaseInstances)
     void MultiDrawElementsInstancedBaseVertexBaseInstanceANGLE(uint32_t mode, IPC::ArrayReferenceTuple<int32_t, int32_t, int32_t, int32_t, uint32_t> countsOffsetsInstanceCountsBaseVerticesAndBaseInstances, uint32_t type)
+#if PLATFORM(COCOA)
+    void CreateEGLSync(MachSendRight syncEvent, uint64_t signalValue) -> (uint64_t sync) Synchronous NotStreamEncodable NotStreamEncodableReply
```
See above, this shouldn't be needed when the function CreateEGLSync is declared in the autogen section.

djg replied:
I had a lot of trouble using the autogen for this function because of `MachSendRight`. I need to add `NotStreamEncodable NotStreamEncodableReply` which I don't think I can do with the autogen methods.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -735,7 +735,19 @@ static bool needsEAGLOnMac()
+        return nullptr;
+    }
+
+    return createEGLSync((id) sharedEvent, signalValue);
```
this could be avoiding C-style cast by using sharedEvent.get()

kkinnunen-apple reviewed (approved):
Nice.

The adoption of more EGLSyncs is a bit problematic since those are per-EGLDisplay, so they're leaked when the context is destroyed unless explicitly tracked. We could address that later.

djg committed:
[WebGL] Abstract working with EGL sync objects created from MTLSharedEvent
https://bugs.webkit.org/show_bug.cgi?id=256954
rdar://problem/109502752

Reviewed by Kimmo Kinnunen.

When the GPU process is enabled, the Web process no long has access to the APIs
required to create shared events passed from the UI process. This change added
APIs to allow creationg of sync objects from events that can be represented by a
MachSendRight (MTLSharedEvent), that can be used in the Web or GPU process.

Support work for WebXR in GPUP.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createEGLSync):
(WebCore::GraphicsContextGLANGLE::destroyEGLSync):
(WebCore::GraphicsContextGLANGLE::clientWaitEGLSyncWithFlush):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::createEGLSync):
(WebCore::GraphicsContextGLCocoa::insertFinishedSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::createSyncWithSharedEvent): Deleted.
(WebCore::GraphicsContextGLCocoa::destroySync): Deleted.
(WebCore::GraphicsContextGLCocoa::clientWaitSyncWithFlush): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::createEGLSync):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(destroyEGLSync):
(clientWaitEGLSyncWithFlush):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createEGLSync):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLSync):
(WebKit::RemoteGraphicsContextGLProxy::clientWaitEGLSyncWithFlush):
* Tools/Scripts/generate-gpup-webgl:

Canonical link: https://commits.webkit.org/264659@main
